packages/block-editor/src/store/reducer.js

```diff
@@ -264,16 +281,34 @@ function withIgnoredBlockChange( reducer ) {
  * @return {Function} Enhanced reducer function.
  */
 const withInnerBlocksRemoveCascade = ( reducer ) => ( state, action ) => {
-	if ( state && action.type === 'REMOVE_BLOCKS' ) {
-		const clientIds = [ ...action.clientIds ];
+	const getAllChildren = ( clientIds ) => {
+		const result = [ ...clientIds ];
 
 		// For each removed client ID, include its inner blocks to remove,
 		// recursing into those so long as inner blocks exist.
-		for ( let i = 0; i < clientIds.length; i++ ) {
-			clientIds.push( ...state.order[ clientIds[ i ] ] );
+		for ( let i = 0; i < result.length; i++ ) {
+			if ( state.order[ result[ i ] ] ) {
+				result.push( ...state.order[ result[ i ] ] );
+			}
 		}
+		return result;
+	};
 
-		action = { ...action, clientIds };
+	if ( state ) {
+		switch ( action.type ) {
+			case 'REMOVE_BLOCKS':
+				action = {
+					...action,
+					removedClientIds: getAllChildren( action.clientIds ),
```

Comment:
Is there a specific reason we need to rename these properties from `clientIds`, rather than just "enhancing" the existing value to account for the cascade? Seems like it might make this a bit simpler to implement, since you wouldn't need the `switch` to vary the behavior.

Reply:
I've been caught off guard in several occasions because of this. The fact that the property don't contain the same value between reducers (selection and blocks), and also the difference between the action creator (which I tend to take a look at to understand the action) led me to this change. 

It is still a different action but instead but I think the fact that we introduce a new key makes it less confusing.

Reply:
I get the motivation for this. It could, however, be pretty jarring for someone who notices the discrepancy and pulls up the definition of `removeBlocks`, since it clearly only returns `clientIds` in its payload.

Perhaps update `removeBlocks` to make that relationship clearer?

```js
// actions.js
yield {
	type: 'REMOVE_BLOCKS',
	clientIds,
	removedClientIds: undefined // Explain bla bla higher-order reducer
};
```

In alternative, if we're really strict about action types, it would be more "correct" to let this higher-order reducer divert from one action type to a new one:

```js
// reducer.js
if ( state ) {
	switch ( action.type ) {
		case 'REMOVE_BLOCKS':
			action = {
				type: DERIVED_TYPE_SIMILAR_TO_REMOVE_BLOCKS,
				removedClientIds: …
```

Reply:
Alternatively, with controls implementation, these days we could probably incorporate this into the action creator itself, rather than as a reducer enhancer. That way the logic is consolidated to one place.